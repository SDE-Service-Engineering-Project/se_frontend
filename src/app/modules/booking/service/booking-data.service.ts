import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Booking } from '../../../models/booking';

export const BOOKING_API = 'http://localhost:8080/api/v1/bookings';

@Injectable({
  providedIn: 'root',
})
export class BookingDataService {
  constructor(private http: HttpClient) {}

  fetchBookings(): Observable<Booking[]> {
    return this.http.get<Booking[]>(BOOKING_API);
  }
}
