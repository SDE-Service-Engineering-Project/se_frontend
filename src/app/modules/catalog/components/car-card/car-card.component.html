<div class="card" style="width: 18rem">
  <img
    src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
    class="card-img-top"
    alt="{{ car.brand }} {{ car.model }}"
  />
  <div class="card-body">
    <h5 class="card-title">{{ car.brand }} {{ car.model }}</h5>
    <p class="card-text">Brand: {{ car.brand }}</p>
    <p class="card-text">Model: {{ car.model }}</p>
    <p class="card-text">Manufacture Year: {{ car.constructionYear }}</p>
    <p class="card-text"><b>Price:</b> {{ car.price | currency: car.currency }} / Day</p>
    <button (click)="selectCar()" style="float: right" class="btn btn-primary">
      Rent
    </button>
  </div>
</div>
