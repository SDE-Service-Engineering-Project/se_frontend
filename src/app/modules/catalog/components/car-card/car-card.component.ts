import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Car } from '../../../../models/car';

@Component({
  selector: 'app-car-card',
  templateUrl: './car-card.component.html',
  styleUrls: ['./car-card.component.sass'],
})
export class CarCardComponent {
  @Input() car: Car = {} as Car;
  @Output() carSelected: EventEmitter<Car> = new EventEmitter<Car>();

  selectCar() {
    this.carSelected.emit(this.car);
  }
}
