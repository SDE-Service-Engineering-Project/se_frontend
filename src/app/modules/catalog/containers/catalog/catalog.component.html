<div class="container" style="margin-top: 1rem">

  <h1 class="mt-2">Explore our arsenal</h1>

  <div class="row">
    <div class="col" *ngFor="let car of cars$ | async">
      <app-car-card style="margin: 1rem" [car]="car" (carSelected)="navigateToCarDetails($event)"></app-car-card>
    </div>

    <div class="container mt-1" *ngIf="(cars$ | async)?.length === 0">
      <div class="row">
        <h1 class="text-center">No Cars found</h1>
      </div>
    </div>
  </div>
</div>
