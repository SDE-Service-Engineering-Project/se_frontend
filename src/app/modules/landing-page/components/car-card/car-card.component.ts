import {Component, EventEmitter, Input, Output} from '@angular/core';
import {Car} from "../../../../models/car";

@Component({
  selector: 'app-car-card',
  templateUrl: './car-card.component.html',
  styleUrls: ['./car-card.component.sass']
})
export class CarCardComponent {
  @Input() car: Car = {} as Car;
  @Output() carRented: EventEmitter<Car> = new EventEmitter<Car>()

  rentCar() {
    this.carRented.emit(this.car);
  }

}
