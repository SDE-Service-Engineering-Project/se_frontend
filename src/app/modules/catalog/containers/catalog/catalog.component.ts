import { Component } from '@angular/core';
import { CarDataService } from '../../services/car-data.service';
import { Observable } from 'rxjs';
import { Car } from '../../../../models/car';
import { ActivatedRoute, Router } from '@angular/router';

@Component({
  selector: 'app-catalog',
  templateUrl: './catalog.component.html',
  styleUrls: ['./catalog.component.sass'],
})
export class CatalogComponent {
  cars$: Observable<Car[]> | undefined;
  from: number | undefined;
  to: number | undefined;

  constructor(
    private carDataService: CarDataService,
    private router: Router,
    private route: ActivatedRoute
  ) {
    this.fetchCars();
  }

  navigateToCarDetails(car: Car) {
    this.router.navigate(['catalog', car.carId]);
  }

  private fetchCars() {
    this.route.queryParams.subscribe((params) => {
      this.from = Date.parse(params['from']);
      this.to = Date.parse(params['to']);
    });

    if (this.from && this.to) {
      this.cars$ = this.carDataService.fetchAvailableCars(this.from, this.to);
    } else {
      this.cars$ = this.carDataService.fetchCars();
    }
  }
}
