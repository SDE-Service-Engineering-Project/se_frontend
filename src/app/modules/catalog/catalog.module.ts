import { NgModule } from '@angular/core';
import { CatalogComponent } from './containers/catalog/catalog.component';
import { CatalogDetailsComponent } from './containers/catalog-details/catalog-details.component';
import { CarCardModule } from './components/car-card/car-card.module';
import { AsyncPipe, CommonModule } from '@angular/common';
import { CatalogRoutingModule } from './config/catalog.routing';

@NgModule({
  declarations: [CatalogComponent, CatalogDetailsComponent],
  imports: [CatalogRoutingModule, CarCardModule, AsyncPipe, CommonModule],
})
export class CatalogModule {}
