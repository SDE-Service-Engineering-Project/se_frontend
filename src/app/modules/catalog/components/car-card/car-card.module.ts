import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CarCardComponent } from './car-card.component';

@NgModule({
  imports: [CommonModule],
  declarations: [CarCardComponent],
  exports: [CarCardComponent],
})
export class CarCardModule {}
