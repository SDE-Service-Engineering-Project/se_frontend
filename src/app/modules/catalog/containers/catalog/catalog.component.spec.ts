import { ComponentFixture } from '@angular/core/testing';

import { CatalogComponent } from './catalog.component';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { CarCardModule } from '../../components/car-card/car-card.module';
import { CarDataService } from '../../services/car-data.service';
import { CarCardComponent } from '../../components/car-card/car-card.component';
import { carMock1, carsMock } from '../../../../utils/testing/mocks/car.mock';
import { Router } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';
import { Component } from '@angular/core';
import { of } from 'rxjs';
import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';

@Component({
  template: '',
})
class RoutedTestComponent {}

describe('CatalogComponent', () => {
  let spectator: Spectator<CatalogComponent>;
  let component: CatalogComponent;
  let fixture: ComponentFixture<CatalogComponent>;
  let navigateSpy: jest.SpyInstance;
  let router: Router;
  let carDataService: CarDataService;
  let httpTestingController: HttpTestingController;

  const createComponent = createComponentFactory({
    component: CatalogComponent,
    imports: [
      CarCardModule,
      HttpClientTestingModule,
      RouterTestingModule.withRoutes([
        { path: '', component: RoutedTestComponent },
      ]),
    ],
    providers: [CarDataService],
  });

  beforeEach(() => {
    spectator = createComponent();
    httpTestingController = spectator.inject(HttpTestingController);
    component = spectator.component;
    fixture = spectator.fixture;
    router = spectator.inject(Router);
    carDataService = spectator.inject(CarDataService);
  });

  beforeEach(() => {
    component.cars$ = of(carsMock);
    spectator.detectChanges();
  });

  it('should render the component', () => {
    expect(fixture.debugElement.nativeElement).toMatchSnapshot();
  });

  it('should navigate to the catalog details page when a car selection is emitted', () => {
    navigateSpy = jest.spyOn(router, 'navigate').mockImplementation();
    getCarCardComponent().carSelected.emit(carMock1);
    expect(navigateSpy).toHaveBeenCalledWith(['catalog', carMock1.carId]);
  });

  function getCarCardComponent(): CarCardComponent {
    return spectator.query(CarCardComponent) as CarCardComponent;
  }
});
