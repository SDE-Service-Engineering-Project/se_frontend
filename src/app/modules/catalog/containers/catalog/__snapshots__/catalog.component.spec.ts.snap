// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogComponent should render the component 1`] = `
<div
  id="root0"
>
  <div
    class="container"
    style="margin-top: 1rem;"
  >
    <h1
      class="mt-2"
    >
      Explore our arsenal
    </h1>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="VW Golf"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                VW Golf
              </h5>
              <p
                class="card-text"
              >
                Brand: VW
              </p>
              <p
                class="card-text"
              >
                Model: Golf
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2020
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €100.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="Opel Kadett"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Opel Kadett
              </h5>
              <p
                class="card-text"
              >
                Brand: Opel
              </p>
              <p
                class="card-text"
              >
                Model: Kadett
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2014
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €200.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="Mercedes A-Class"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Mercedes A-Class
              </h5>
              <p
                class="card-text"
              >
                Brand: Mercedes
              </p>
              <p
                class="card-text"
              >
                Model: A-Class
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2018
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €300.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="VW Golf"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                VW Golf
              </h5>
              <p
                class="card-text"
              >
                Brand: VW
              </p>
              <p
                class="card-text"
              >
                Model: Golf
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2020
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €100.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="Opel Kadett"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Opel Kadett
              </h5>
              <p
                class="card-text"
              >
                Brand: Opel
              </p>
              <p
                class="card-text"
              >
                Model: Kadett
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2014
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €200.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="Mercedes A-Class"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Mercedes A-Class
              </h5>
              <p
                class="card-text"
              >
                Brand: Mercedes
              </p>
              <p
                class="card-text"
              >
                Model: A-Class
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2018
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €300.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="VW Golf"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                VW Golf
              </h5>
              <p
                class="card-text"
              >
                Brand: VW
              </p>
              <p
                class="card-text"
              >
                Model: Golf
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2020
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €100.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="Opel Kadett"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Opel Kadett
              </h5>
              <p
                class="card-text"
              >
                Brand: Opel
              </p>
              <p
                class="card-text"
              >
                Model: Kadett
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2014
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €200.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="Mercedes A-Class"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Mercedes A-Class
              </h5>
              <p
                class="card-text"
              >
                Brand: Mercedes
              </p>
              <p
                class="card-text"
              >
                Model: A-Class
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2018
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €300.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="VW Golf"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                VW Golf
              </h5>
              <p
                class="card-text"
              >
                Brand: VW
              </p>
              <p
                class="card-text"
              >
                Model: Golf
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2020
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €100.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="Opel Kadett"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Opel Kadett
              </h5>
              <p
                class="card-text"
              >
                Brand: Opel
              </p>
              <p
                class="card-text"
              >
                Model: Kadett
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2014
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €200.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      <div
        class="col"
      >
        <app-car-card
          style="margin: 1rem;"
        >
          <div
            class="card"
            style="width: 18rem;"
          >
            <img
              alt="Mercedes A-Class"
              class="card-img-top"
              src="https://pictures.dealer.com/c/costamesacadillac/0679/a976281f56ff06d46dc390283aa14da6x.jpg?impolicy=downsize_bkpt&imdensity=1&w=520"
            />
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Mercedes A-Class
              </h5>
              <p
                class="card-text"
              >
                Brand: Mercedes
              </p>
              <p
                class="card-text"
              >
                Model: A-Class
              </p>
              <p
                class="card-text"
              >
                Manufacture Year: 2018
              </p>
              <p
                class="card-text"
              >
                <b>
                  Price:
                </b>
                 €300.00 / Day
              </p>
              <button
                class="btn btn-primary"
                style="float: right;"
              >
                 Rent 
              </button>
            </div>
          </div>
        </app-car-card>
      </div>
      
      
    </div>
  </div>
</div>
`;
